src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Contracts/ITelemetryReporter.cs

```diff
@@ -2,15 +2,13 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Immutable;
-
 namespace Microsoft.CodeAnalysis.Contracts.Telemetry;
 
 internal interface ITelemetryReporter : IDisposable
 {
     void InitializeSession(string telemetryLevel, string? sessionId, bool isDefaultSession);
-    void Log(string name, ImmutableDictionary<string, object?> properties);
+    void Log(string name, ReadOnlyMemory<Property> properties);
```

Comment:
Very open to different ideas on what to use here (ReadOnlyMemory, IEnumerable, List). List felt a bit like it might be exposing too much, but maybe it's ok? The part I don't care for with using ReadOnlyMemory is that it didn't allow me to use a List in the cached pool. By using an array, I don't get it cleared when the pooled object is freed, which means the Property instances can be held onto for an unbounded amount of time.

Reply:
Iteration 2 is the view of what this looks like using a List. (I suppose IReadOnlyList would be very similar, but with a cleaner abstraction being used but with the small perf implications of using an interface)

Reply:
Iteration 3: With KeyValuePair instead per PR suggestions. Going to go ahead with this unless there is feedback otherwise.